clients/stellar-relay-lib/src/connection/connector/message_reader.rs

```diff
@@ -43,39 +46,58 @@ pub(crate) async fn poll_messages_from_stellar(
 		}
 
 		// check for messages from Stellar Node.
-		let xdr = match read_message_from_stellar(&mut connector).await {
-			Err(e) => {
+		// if reading took too much time, flag it as "disconnected"
+		let xdr = match timeout(
+			Duration::from_secs(READ_TIMEOUT_IN_SECS),
+			read_message_from_stellar(&mut connector),
+		)
+		.await
+		{
+			Ok(Ok(xdr)) => xdr,
+			Ok(Err(e)) => {
 				error!("poll_messages_from_stellar(): {e:?}");
-				break;
+				break
+			},
+			Err(_) => {
+				error!("poll_messages_from_stellar(): timed out");
+				break
 			},
-			Ok(xdr) => xdr,
 		};
 
 		match connector.process_raw_message(xdr).await {
-			Ok(Some(stellar_msg)) =>
-			// push message to user
-			{
-				if let Err(e) = send_to_user_sender.send(stellar_msg.clone()).await {
+			Ok(Some(stellar_msg)) => {
+				// push message to user
+				let stellar_msg_as_base64_xdr = stellar_msg.to_base64_xdr();
+				if let Err(e) = send_to_user_sender.send(stellar_msg).await {
 					warn!("poll_messages_from_stellar(): Error occurred during sending message {} to user: {e:?}",
-						String::from_utf8(stellar_msg.to_base64_xdr())
-						.unwrap_or_else(|_| format!("{:?}", stellar_msg.to_base64_xdr()))
-					);
+					String::from_utf8(stellar_msg_as_base64_xdr.clone())
+					.unwrap_or_else(|_| format!("{stellar_msg_as_base64_xdr:?}"))
+				);
 				}
+				tokio::task::yield_now().await;
 			},
-			Ok(None) => {},
+			Ok(None) => tokio::task::yield_now().await,
 			Err(e) => {
 				error!("poll_messages_from_stellar(): Error occurred during processing xdr message: {e:?}");
 				break;
 			},
 		}
 	}
 
+	// push error to user
+	if let Err(e) = send_to_user_sender.send(crate_specific_error()).await {
```

Comment:
Something happened inside here, and it must be communicated back to the caller/user.